design/extension.go

```diff
@@ -0,0 +1,963 @@
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("kfserving-001")
+			})
+			Field(2, "service_id", String, "Extension service identifier", func() {
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("s3")
+			})
+			Required("extension_id", "service_id")
+		})
+
+		Result(ArrayOf(ExtensionCredentials), "Return all credentials associated with an extension service.")
+
+		HTTP(func() {
+			GET("/extensions/{extension_id}/services/{service_id}/credentials")
+			Response(StatusOK)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+		})
+	})
+
+	Method("updateCredentials", func() {
+		Description("Update a set of credentials belonging to an extension service registered in FuseML")
+
+		Payload(ExtensionCredentials, "Extension credentials update request")
+
+		Result(ExtensionCredentials, "Return the updated set of credentials.")
+
+		Error("BadRequest", func() {
+			Description("If the set of extension credentials does not have the required fields, should return 400 Bad Request.")
+		})
+
+		Error("NotFound", func() {
+			Description("If the extension, the service or the set of credentials are not found, should return 404 Not Found.")
+		})
+
+		HTTP(func() {
+			PUT("/extensions/{extension_id}/services/{service_id}/credentials/{id}")
+			Response(StatusOK)
+			Response("NotFound", StatusNotFound)
+			Response("BadRequest", StatusBadRequest)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+	Method("deleteCredentials", func() {
+		Description("Delete a set of extension credentials")
+
+		Payload(func() {
+			Field(1, "extension_id", String, "Extension identifier", func() {
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("kfserving-001")
+			})
+			Field(2, "service_id", String, "Extension service identifier", func() {
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("s3")
+			})
+			Field(3, "id", String, "Uniquely identifies a set of credentials within the scope of an extension service", func() {
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("cred-user-12bb")
+			})
+			Required("extension_id", "service_id", "id")
+		})
+
+		Error("NotFound", func() {
+			Description("If there is no extension, service or set of credentials with the given ID, should return 404 Not Found.")
+		})
+
+		Error("BadRequest", func() {
+			Description("If the set of extension credentials cannot be deleted, should return 400 Bad Request.")
+		})
+
+		HTTP(func() {
+			DELETE("/extensions/{extension_id}/services/{service_id}/credentials/{id}")
+			Response(StatusNoContent)
+			Response("NotFound", StatusNotFound)
+			Response("BadRequest", StatusBadRequest)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+})
+
+// Extension descriptor
+var Extension = Type("Extension", func() {
+	tag := 1
```

Comment:
eh, what's the tag for? I see it for the first time..

Reply:
gRPC fields require a unique tag for each attribute